src/electron/process_manager.ts

```diff
@@ -316,56 +236,119 @@ class ChildProcessHelper {
+    this.process = new ChildProcessHelper(pathToEmbeddedBinary('go-tun2socks', 'tun2socks'));
   }
 
-  start(config: cordova.plugins.outline.ServerConfig) {
-    // ss-local -s x.x.x.x -p 65336 -k mypassword -m aes-128-cfb -l 1081 -u
-    const args = ['-l', this.proxyPort.toString()];
-    args.push('-s', config.host || '');
-    args.push('-p', '' + config.port);
-    args.push('-k', config.password || '');
-    args.push('-m', config.method || '');
-    args.push('-u');
+  async start(checkConnectivity: boolean) {
+    if (this.process.isRunning) {
+      return this.restart(checkConnectivity);
+    }
 
-    this.launch(args);
+    this.process.launch(this.getProcessArgs(checkConnectivity));
+
+    return new Promise((resolve, reject) => {
+      // Declare success when tun2socks is running.
+      this.process.onStdout = (data?: string | Buffer) => {
+        if (data && data.toString().includes('tun2socks running')) {
+          this.process.onStdout = undefined;
+          this.process.onExit = this.exitListener;
+          if (isWindows) {
+            powerMonitor.on('suspend', this.suspendListener.bind(this));
+            powerMonitor.on('resume', this.resumeListener.bind((this)));
+          }
+          resolve();
+        }
+      };
+
+      // Wait for an early exit due to connectvity failures.
+      this.process.onExit = (code?: number, signal?: string) => {
+        console.log('tun2socks exited with code', code);
+        if (this.exitListener) {
+          this.exitListener();
+        }
+        // code === 0 should not happen unless invoked with `-version`;
+        // treat it like an unexpected error.
+        reject(errors.fromErrorCode(code || errors.ErrorCode.UNEXPECTED));
+      };
+    });
   }
-}
 
-class Tun2socks extends ChildProcessHelper {
-  constructor(private proxyAddress: string, private proxyPort: number) {
-    super(pathToEmbeddedBinary('go-tun2socks', 'tun2socks'));
+  stop() {
+    if (isWindows) {
+      powerMonitor.removeListener('suspend', this.suspendListener.bind(this));
+      powerMonitor.removeListener('resume', this.resumeListener.bind(this));
+    }
+    this.process.stop();
+  }
+
+  private async restart(checkConnectivity: boolean) {
+    // Swap out the current listener, restart once the current process exits.
+    return new Promise(resolve => {
+      this.process.onExit = async () => {
+        console.log('restarting tun2socks');
+        await this.start(checkConnectivity);
+        resolve();
+      };
+      this.stop();
+    });
   }
 
-  start(isUdpEnabled: boolean) {
-    // ./tun2socks \
-    //   -tunName "tap0901:outline-tap0:10.0.85.2:10.0.85.0:255.255.255.0" \
+  private suspendListener() {
+    console.log('system suspending');
+    // Windows: when the system suspends, tun2socks terminates due to the TAP device getting closed.
+    // Swap out the current listener, restart once the system resumes.
+    this.process.onExit = () => {
```

Comment:
I think there's a race condition here, in this scenario:
1. Tun2socks.suspendListener() runs due to a suspend event and overwrites this.process.onExit
2. ConnectionManager.stop() -> Tun2socks.stop() -> ChildProcessHelper.stop() runs due to a user action
3. Tun2socks.resumeListener() runs

In this case, ConnectionManager.onceStopped will never resolve, because it missed the event.

It's possible that Node provides some guarantee about the timing of suspend and resume events, but I would be surprised.  Usually those kinds of events are pretty sloppy, occurring a few seconds before suspend and a few seconds after resume.

There might be a more elegant solution if `Tun2socks` handles the network change events itself, but I'm not entirely sure.

Reply:
While the race is certainly possible, it seems hard to get into that state as I was not able to reproduce it. If it's an actual concern, I think a larger refactor/redesign would be necessary.

> There might be a more elegant solution if Tun2socks handles the network change events itself, but I'm not entirely sure.

I'm not sure either. `Tun2socks` would have to know about the routing service, which is not ideal. I've been thinking a solution would be to expose an option in the binary to only perform the connectivity checks. This way we wouldn't have to restart tun2socks on every network check and we could contain (the TAP-processing) tun2socks lifecycle in `Tun2socks`. Wdyt?